docs/latest/design/extending-envoy-gateway.md

```diff
@@ -0,0 +1,302 @@
+real way to prevent one extension from overwriting or breaking modifications to xDS resources that were made by another extension that was executed first.
+
+## Overview
+
+Envoy Gateway can be extended by vendors by means of an extension server developed by the vendor and deployed alongside Envoy Gateway.
+An extension server can make use of one or more pre/post hooks inside Envoy Gateway before and after its major components (translator, etc.) to allow the extension to modify the data going into or coming out of these components.
+An extension can be created external to Envoy Gateway as its own Kubernetes deployment or loaded as a sidecar. gRPC is used for the calls between Envoy Gateway and an extension. In the hook call, Envoy Gateway sends data as well 
+as context information to the extension and expects a reply with a modified version of the data that was sent to the extension. Since extensions fundamentally alter the logic and data that Envoy Gateway provides, Extension projects assume responsibility for any bugs and issues 
+they create as a direct result of their modification of Envoy Gateway.
+
+## Diagram
+
+![Architecture](../images/extension-example.png)
+
+## Registering Extensions in Envoy Gateway
+
+Information about the extension that Envoy Gateway needs to load is configured in the Envoy Gateway config. The extension must specify:
+
+- Which API groups the extension supports
+- The `Kind`s supported under each API group
+- Extension service connection info (host, port, TLS settings, etc.)
+
+An example configuration:
+
+```yaml
+apiVersion: config.gateway.envoyproxy.io/v1alpha1
+kind: EnvoyGateway
+extension:
+  apiGroups:
+  - apiGroup: example.myextension.io
+    apiVersion: v2
+    kinds:
+    - OAuth2Filter
+  service:
+    host: my-extension.example
+    port: 443
+    tls:
+      type: secret
+      secret:
+        name: my-extension
+        namespace: example
+```
+
+This configuration is required to be provided at bootstrap and modifying the registered extension during runtime is not currently supported.
+Envoy Gateway will keep track of the registered extension and its `apiGroups` and `kinds` when processing Gateway API resources.
+
+## Extending Gateway API and the Data Plane
+
+Envoy Gateway manages [Envoy][] deployments, which act as the data plane that handles actual user traffic. Users configure the data plane using the K8s Gateway API resources which Envoy 
+Gateway converts into [Envoy specific configuration (xDS)][] to send over to Envoy.
+
+Gateway API offers [ExtensionRef filters][] and [Policy Attachments][] as extension points for implementers to use. Envoy Gateway extends the Gateway API using these extension points to provide support for [rate limiting][] 
```

Comment:
how can we better frame this to highlight 2 types of extensions
* Native to the project
* Extending the control plane by the vendor/extension dev